### PR TITLE
Add gltf import support

### PR DESCRIPTION
**Pull request by beverlylytle**
### What type of change is this?

- [x] New feature in a **backwards-compatible** manner.

Adds functionality to import the tree structure and data of a glTF file, including the attached meshes.

### CODE REVIEW

File: src/compas/files/gltf.py
Comment (tomvanmele) [inline src/compas/files/gltf.py]: ?

Comment (beverlylytle) [reply src/compas/files/gltf.py]: I could clarify with "Usage of MultiMeshViewer unclear _to me_".  This seems to follow the pattern established in stl.py, but when I tried to run stl.py and gltf.py, it didn't work.  I left this in here so that some one might explain viewers to me (but would be removed/fixed before merging).

Comment (tomvanmele) [reply src/compas/files/gltf.py]: okay, nor problem. do you have `compas_viewers` installed?

Comment (beverlylytle) [reply src/compas/files/gltf.py]: yes

Comment (tomvanmele) [reply src/compas/files/gltf.py]: also, in py3, `keys` and `values` are iterators by default so you may want to convert the output to an actual list
```python
viewer.meshes = list(meshes.values())
```

Comment (tomvanmele) [reply src/compas/files/gltf.py]: https://compas-dev.github.io/examples/mesh-subd-comparison/mesh-subd-comparsion.html

Comment (beverlylytle) [reply src/compas/files/gltf.py]: It seems that the import statement in stl.py was outdated and should be replaced with 
    `from compas_viewers.meshviewer import MeshViewer`
This is what was causing me confusion.

**Comment by beverlylytle** [comment]
This PR will be updated substantially some time tomorrow.  Please hold off reviewing until then.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: Totally minor comment, but perhaps best to just `import math` and use `math.fsum` (I didn't recognize it as a math module function when I saw it used)

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: not sure what sphinx will do with this blank line, it might think the list of attributes is over and the rest is description.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]:
```suggestion
        Dictionary of (node_key, :class:`GLTFNode`) pairs.
```
Here we can use the class directive thingy to have these linked/clickable on the doc.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py:195]: Maybe wrap all this block until line ~206 in a `try..finally`, and call `self._content.release()` before nullifying the variable in the `finally` block to make sure we don't keep any dangling reference.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: Why is length suffixed with an underscore? Some kind of reserved keyword?

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py:239]: If that is not true, should we raise an exception or is it ok to just ignore?

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: Wouldn't it be safer to keep this check as a string, because in the future, a theoretical patch version (eg. `2.0.1`) would break the float cast.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: These component types (`5120, 5121, 5122, 5123`) would be a better if we declare as constants at the top, maybe changing the `COMPONENT_TYPE_ENUM` to use it, something like:
```python
COMPONENT_TYPE_BYTE = 5120
COMPONENT_TYPE_UBYTE = 5121
..
COMPONENT_TYPE_ENUM = {
    COMPONENT_TYPE_BYTE: 'b',  # BYTE
    COMPONENT_TYPE_UBYTE: 'B',  # UNSIGNED_BYTE
    ..
}
```

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]:
```suggestion
    reader : :class:`GLTFReader`
```

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]:
```suggestion
    scenes : list of :class:`GLTFScene`
```

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: There's a way to directly link to the method, but I don't remember exactly the syntax, is something like :attr:`GLTFScene.nodes`...

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py:392]: These buffers are tricky to release correctly, perhaps the `GLTFReader` needs to have a `release` method itself, that is called once reading is completed. However, I'm not sure if reading is completed in our context or in the user's. If that is the case, maybe the `GLTF` class itself needs a way of releasing, and the user is in charge of calling it to trigger all deallocations.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: Shouldn't we keep edges in a dictionary? I assume it's now a list to make it easy to create a network afterward, but if used standalone, it would be faster to use dictionary lookups instead of iterating over a list.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py]: That's .. intense ;)  I don't know if it can be simplified for readability, maybe if you assign the lambdas to a variable before, then it can be more "humanly parsable" 😄

**Review by gonzalocasas — APPROVED** [review]
LGTM

Looks excellent! I added a few comments, but overall, it's great. I have not tested it thou, but since I know you are still planning on making a few more changes for IronPython compat, I'll test after that! ;)

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py]: I warned you... Worst line of code ever.

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py]: I'm a little confused as to what you have in mind for this dictionary.  Do you mean something of the form {parent_key: <list of children's keys>}? (This sort of thing is already recoverable from the nodes.) Or do you mean that the edge tuples should be stored as a set rather than a list?  I am almost tempted to delete this attribute completely because I already expect the user to create the appropriate dictionary for the vertices.  Why wouldn't I also expect this for the edges?

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py]: ok

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py]: Clearly I was faking it went it came to anything sphinx related...

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py]: :meth:GLTFScene.nodes ?

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py:239]: This is checking whether there is a second chunk of binary data.  It may be that everything is stored as a data uri and so this chunk wouldn't be there.

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py]: or :func: ?

File: src/compas/files/gltf.py
Comment (beverlylytle) [reply src/compas/files/gltf.py:195]: What about appending the following?
```
try:
    self._content.release()
except AttributeError:
    pass
finally:
   self._content = None
```
Or is there another exception you are thinking that should be caught earlier within this block?

File: src/compas/files/gltf.py
Comment (tetov) [reply src/compas/files/gltf.py]: :meth

(Had to look this up recently. Here’s docs re hyperlinking in Sphinx for python code: https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects)

File: src/compas/files/gltf.py
Comment (beverlylytle) [inline src/compas/files/gltf.py:369]: Slicing the memoryview and calling tobytes() is the change I had to make for IronPython.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [reply src/compas/files/gltf.py:195]: I didn't mean so much the case for exception handling as the disposal/release of the buffer, but I see you added a few methods to release buffers, so I'll mark this as resolved.

File: src/compas/files/gltf.py
Comment (gonzalocasas) [reply src/compas/files/gltf.py:239]: 👍

File: src/compas/files/gltf.py
Comment (gonzalocasas) [reply src/compas/files/gltf.py:369]: 👍

File: src/compas/files/gltf.py
Comment (gonzalocasas) [inline src/compas/files/gltf.py:573]: 👍 much easier to digest!

File: src/compas/files/gltf.py
Comment (gonzalocasas) [reply src/compas/files/gltf.py]: but `nodes` is an attribute, not a method, isn't it?

Comment (beverlylytle) [reply src/compas/files/gltf.py]: Yes, I've used :attr:.

**Pull request merged**
